node/authorship/src/block_builder.rs

```diff
@@ -0,0 +1,89 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use pallet_gear_rpc_runtime_api::GearApi as GearRuntimeApi;
+use sc_block_builder::{BlockBuilder, BlockBuilderApi, BuiltBlock};
+use sc_client_api::backend;
+use sp_api::{ApiExt, ApiRef, ProvideRuntimeApi, TransactionOutcome};
+use sp_blockchain::Error;
+use sp_core::ExecutionContext;
+use sp_runtime::{generic::BlockId, traits::Block as BlockT};
+
+pub struct BlockBuilderExt<'a, Block: BlockT, A: ProvideRuntimeApi<Block>, B> {
+    block_builder: BlockBuilder<'a, Block, A, B>,
+    api: ApiRef<'a, A::Api>,
+    parent_hash: Block::Hash,
+}
+
+impl<'a, Block, A, B> BlockBuilderExt<'a, Block, A, B>
+where
+    Block: BlockT,
+    A: ProvideRuntimeApi<Block> + 'a,
+    A::Api: ApiExt<Block, StateBackend = backend::StateBackendFor<B, Block>>
+        + BlockBuilderApi<Block>
+        + GearRuntimeApi<Block>,
+    B: backend::Backend<Block>,
+{
+    /// Creating a block builder by wrapping an sc_block_builder::BlockBuilder object
+    pub fn new(
+        block_builder: BlockBuilder<'a, Block, A, B>,
+        api: ApiRef<'a, A::Api>,
+        parent_hash: Block::Hash,
+    ) -> Self {
+        Self {
+            block_builder,
+            api,
+            parent_hash,
+        }
+    }
+
+    /// Push onto the block's list of extrinsics.
+    pub fn push(&mut self, xt: <Block as BlockT>::Extrinsic) -> Result<(), Error> {
+        self.block_builder.push(xt)
+    }
+
+    /// Consume the builder to build a valid `Block` containing all pushed extrinsics.
+    pub fn build(self) -> Result<BuiltBlock<Block, backend::StateBackendFor<B, Block>>, Error> {
+        self.block_builder.build()
+    }
+
+    /// Create the inherents for the block.
+    pub fn create_inherents(
+        &mut self,
+        inherent_data: sp_inherents::InherentData,
+    ) -> Result<Vec<Block::Extrinsic>, Error> {
+        self.block_builder.create_inherents(inherent_data)
+    }
+
+    /// Estimate the size of the block in the current state.
+    pub fn estimate_block_size(&self, include_proof: bool) -> usize {
+        self.block_builder.estimate_block_size(include_proof)
+    }
+
+    pub fn create_terminal_extrinsic(&mut self) -> Result<Block::Extrinsic, Error> {
+        let block_id = BlockId::Hash(self.parent_hash);
+        self.api
+            .execute_in_transaction(move |api| {
+                TransactionOutcome::Rollback(api.run_queue_extrinsic_with_context(
```

Comment:
can't find from where `_with_context` method appeared?

Reply:
It's some kind of Substrate magic. Didn't find it either but yet all the runtime api methods called from transactional environment have it.

Reply:
Seems like extending for rpc methods to have bn stamp as an argument, okay